src/main/java/com/github/mizool/jediscache/JedisCache.java

```diff
@@ -0,0 +1,353 @@
+            converter
+            = new JedisCacheJavaSerializationConverter<>(configuration.getKeyType(), configuration.getValueType());
+        keyConverter = converter;
+        valueConverter = converter;
+        jedisCacheName = name.getBytes(Charsets.UTF_8);
+
+        closed = new AtomicBoolean(false);
+    }
+
+    @Override
+    public V get(K key)
+    {
+        verifyOpen();
+        try (Jedis jedis = jedisPool.getResource())
+        {
+            return valueConverter.toValue(jedis.hget(jedisCacheName, keyConverter.fromKey(key)));
+        }
+    }
+
+    @Override
+    public Map<K, V> getAll(Set<? extends K> keys)
+    {
+        verifyOpen();
+        Map<byte[], byte[]> allEntries;
+        try (Jedis jedis = jedisPool.getResource())
+        {
+            allEntries = jedis.hgetAll(jedisCacheName);
+        }
+        ImmutableMap.Builder<K, V> builder = ImmutableMap.builder();
+        keys.forEach(key -> {
+            byte[] serializedKey = keyConverter.fromKey(key);
+            if (allEntries.containsKey(serializedKey))
+            {
+                builder.put(key, valueConverter.toValue(allEntries.get(serializedKey)));
+            }
+        });
+        return builder.build();
+    }
+
+    @Override
+    public boolean containsKey(K key)
+    {
+        verifyOpen();
+        try (Jedis jedis = jedisPool.getResource())
+        {
+            return jedis.hexists(jedisCacheName, keyConverter.fromKey(key));
+        }
+    }
+
+    @Override
+    public void loadAll(
+        Set<? extends K> keys, boolean replaceExistingValues, CompletionListener completionListener)
+    {
+        throw new UnsupportedOperationException("listener support is not implemented");
+    }
+
+    @Override
+    public void put(K key, V value)
+    {
+        verifyOpen();
+        try (Jedis jedis = jedisPool.getResource())
+        {
+            jedis.hset(jedisCacheName, keyConverter.fromKey(key), valueConverter.fromValue(value));
+        }
+    }
+
+    @Override
+    public V getAndPut(K key, V value)
+    {
+        verifyOpen();
+        V result = get(key);
+        put(key, value);
+        return result;
+    }
+
+    @Override
+    public void putAll(Map<? extends K, ? extends V> map)
+    {
+        map.forEach(this::put);
+    }
+
+    @Override
+    public boolean putIfAbsent(K key, V value)
+    {
+        verifyOpen();
+        try (Jedis jedis = jedisPool.getResource())
+        {
+            return jedis.hsetnx(jedisCacheName, keyConverter.fromKey(key), valueConverter.fromValue(value)) == 1;
+        }
+    }
+
+    @Override
+    public boolean remove(K key)
+    {
+        verifyOpen();
+        try (Jedis jedis = jedisPool.getResource())
+        {
+            return jedis.hdel(jedisCacheName, keyConverter.fromKey(key)) == 1;
+        }
+    }
+
+    @Override
+    public boolean remove(K key, V oldValue)
+    {
+        return get(key).equals(oldValue) && remove(key);
```

Comment:
there's a race condition here:

1. retrieve the current value from redis
2. compare it to the given value
3. send the new value to redis

At some point between 1 and 3, somebody else can send a different value to redis.

The only way to fix this I can think of would be a "compare and set" operation inside Redis (comparable to Cassandra's lightweight transactions). Unfortunately, Redis hash structures do not seem to offer such an operation.

The least we can do is to make a note about this scenario, both in the source code and in the project README (probably in a "Limitations" or "Known issues" section).

Reply:
Note that most of the overridden methods from the cache api talk about atomicity, which we can't guarantee due to redis' limitations. I noticed that during my review but did not think about having a "Limitations" section in the readme.

Reply:
@schroenser Actually, since Redis is single-threaded, and uses queues for incoming commands, its operations are actually quite atomic.

Reply:
A single Redis operation may be atomic, but if we invoke two redis operations for one JCache operation, the latter is not at all atomic. So we definitively need the README section.